src/main/java/ie/gmit/softwareeng/labexam/Playlist.java

```diff
@@ -1,71 +1,95 @@
 package ie.gmit.softwareeng.labexam;
 
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
 public class Playlist {
+
+    private String playlistName;
+    private List<Song> songList = new ArrayList<>();
+    private int currentSong = 0;
+
     public Playlist(String playlistName) {
-        throw new UnsupportedOperationException("Not implemented yet");
+        this.playlistName=playlistName;
     }
 
     public String getName() {
-        throw new UnsupportedOperationException("Not implemented yet");
+        return playlistName;
     }
 
     public boolean contains(Song song) {
-        throw new UnsupportedOperationException("Not implemented yet");
+        return songList.contains(song);
     }
 
     public void addSong(Song song) {
-        throw new UnsupportedOperationException("Not implemented yet");
+        if (songList.contains(song)) throw new PlaylistException("Song already exists in the playlist");
+       songList.add(song);
     }
 
     public void removeSong(Song song) {
-        throw new UnsupportedOperationException("Not implemented yet");
+        if (!songList.contains(song)) throw new PlaylistException("Song does not exists in the playlist");
+        songList.remove(song);
     }
 
     public int getNumberOfSongs() {
-        throw new UnsupportedOperationException("Not implemented yet");
+
+        return songList.size();
     }
 
     public Song getSongAt(int index) {
-        throw new UnsupportedOperationException("Not implemented yet");
+        return songList.get(index);
     }
 
     public void clear() {
-        throw new UnsupportedOperationException("Not implemented yet");
+        songList.clear();
     }
 
     public void reverse() {
-        throw new UnsupportedOperationException("Not implemented yet");
+        Collections.reverse(songList);
     }
 
     public Song[] getAllSongs() {
-        throw new UnsupportedOperationException("Not implemented yet");
+        Song[] songArray = new Song[songList.size()];
+
+        for (int i = 0; i < songList.size(); i++) {
```

Comment:
you could have used something like this to save you looping through the list
```
Song[] songArray = new Song[songs.size()];
return songs.toArray(songArray);
```
This is fine too though